waspc/cli/src/Wasp/Cli/GithubRepo.hs

```diff
@@ -0,0 +1,129 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Wasp.Cli.GithubRepo where
+
+import qualified Codec.Archive.Tar as Tar
+import qualified Codec.Compression.GZip as GZip
+import Control.Exception (SomeException, try)
+import Data.Aeson
+  ( FromJSON,
+    parseJSON,
+    withObject,
+    (.:),
+  )
+import qualified Data.ByteString.Lazy as BL
+import Data.Functor ((<&>))
+import Data.List (intercalate)
+import Data.Maybe (fromJust, maybeToList)
+import Network.HTTP.Conduit (simpleHttp)
+import qualified Network.HTTP.Simple as HTTP
+import Path.IO (copyDirRecur, removeDirRecur)
+import StrongPath (Abs, Dir, File, Path', (</>))
+import qualified StrongPath as SP
+import StrongPath.Path (toPathAbsDir)
+import System.Directory (createDirectoryIfMissing)
+import Wasp.Cli.FileSystem (getUserCacheDir)
+
+type GithubRepo = (GithubRepoOwner, GithubRepoName)
+
+type GithubRepoOwner = String
+
+type GithubRepoName = String
+
+fetchFolderFromGithubRepoToDisk ::
+  forall d.
+  GithubRepo ->
+  String ->
+  Path' Abs (Dir d) ->
+  IO (Either String ())
+fetchFolderFromGithubRepoToDisk (githubOrg, repoName) pathToFolderInRepo destinationOnDisk = do
+  try
+    (withTempDir ".wasp-templates-temp" downloadFolderToDisk)
+    <&> either showException Right
+  where
+    downloadFolderToDisk :: Path' Abs (Dir d) -> IO ()
+    downloadFolderToDisk tempDir = do
+      downloadFile githubArchiveUrl tempDir archiveFileName >>= unpackArchive tempDir
+      copyRepoFolderToDestination tempDir pathToFolderInRepo destinationOnDisk
+      where
+        githubArchiveUrl = intercalate "/" ["https://github.com", githubOrg, repoName, "archive", archiveFileName]
+        archiveFileName = repoReference ++ ".tar.gz"
+
+    -- Which point in repo history to download (a branch or commit hash)
+    repoReference :: String
+    repoReference = "main"
+
+    downloadFile :: String -> Path' Abs (Dir d) -> String -> IO (Path' Abs (File f))
+    downloadFile downloadUrl destinationDir fileName = do
+      let destinationPath = destinationDir </> fromJust (SP.parseRelFile fileName)
+      simpleHttp downloadUrl >>= BL.writeFile (SP.fromAbsFile destinationPath)
+      return destinationPath
+
+    unpackArchive :: Path' Abs (Dir d) -> Path' Abs (File f) -> IO ()
+    unpackArchive destinationDir pathToTheTarFile = do
+      Tar.unpack (SP.fromAbsDir destinationDir) . Tar.read . GZip.decompress =<< BL.readFile (SP.fromAbsFile pathToTheTarFile)
+
+    copyRepoFolderToDestination :: Path' Abs (Dir d) -> String -> Path' Abs (Dir d) -> IO ()
+    copyRepoFolderToDestination archiveRootDir pathToFolderInRepo' destinationOnDisk' = do
+      copyDirRecur (toPathAbsDir pathToDownloadededFolder) (toPathAbsDir destinationOnDisk')
+      where
+        pathToDownloadededFolder = archiveRootDir </> rootArchiveFolder </> fromJust (SP.parseRelDir pathToFolderInRepo')
+        -- Github repo tars have a root folder that is named after the repo name and the reference (branch or tag).
```

Comment:
Ok, this is knowledge specific to the fact that the repo folder was obtained from unarchiving a GH repo archive.
But function itself in its name doesn't specify that assumption. it does use `archiveRootDir` for variable name so that helps a tiny bit, but not enough I think.

Idea: rename this function to `getPathToFolderInGhArchive` or smth like that and it should return path, and then when you need it, you can directly call `copyDirRecur` on it.

Or go a bit further and make these functions you have in `where` more prominent and better named and again, where you need this, directly use those + `copyDirRecur`.

My point is that this function probably shouldn't exist, it is too specific and assumes a lot and is hard to name therefore, it is better to break into smaller pieces, make sure those are named very well, and compose them as needed and when needed. That should result with the most readable code + it will be composable.